chain/network/src/accounts_data/mod.rs

```diff
@@ -0,0 +1,240 @@
+        // Filter out non-interesting data, so that we never check signatures for valid non-interesting data.
+        // Bad peers may force us to check signatures for fake data anyway, but we will ban them after first invalid signature.
+        // It locks epochs for reading for a short period.
+        let mut found = HashSet::new();
+        let mut data_and_keys = vec![];
+        let epochs = self.epochs.read();
+        for d in data {
+            // We want the communication needed for broadcasting per-account data to be minimal.
+            // Therefore broadcasting multiple datasets per account is considered malicious
+            // behavior, since all but one are obviously outdated.
+            let data_id = (d.epoch_id.clone(), d.account_id.clone());
+            if found.contains(&data_id) {
+                return (vec![], Some(Error::SingleAccountMultipleData));
+            }
+            found.insert(data_id);
+            // There is a limit on the amount of RAM occupied by per-account datasets.
+            // Broadcasting larger datasets is considered malicious behavior.
+            if d.payload().len() > network_protocol::MAX_ACCOUNT_DATA_SIZE_BYTES {
+                return (vec![], Some(Error::DataTooLarge));
+            }
+            // It is fine to broadcast account data for epochs that we don't care about.
+            let epoch = match epochs.0.get(&d.epoch_id) {
+                Some(e) => e,
+                None => continue,
+            };
+            // It is NOT fine to broadcast account data which is not relevant to the given epoch.
+            let account = match epoch.0.get(&d.account_id) {
+                Some(a) => a,
+                None => return (vec![], Some(Error::InvalidAccount)),
+            };
+            // It is fine to broadcast data we already know about.
+            if !account.is_new(d.timestamp) {
+                continue;
+            }
+            data_and_keys.push((d, account.key.clone()));
+        }
+        drop(epochs);
+
+        // We verify signatures synchronously for now.
+        // To verify signatures in parallel, we should have a way to stop verification at the first invalid one.
+        let mut data = vec![];
+        for (d, key) in data_and_keys {
+            if !d.payload().verify(&key) {
+                return (data, Some(Error::InvalidSignature));
+            }
+            data.push(d);
+        }
+        return (data, None);
+    }
+
+    /// Verifies the signatures and inserts verified data to the cache.
+    /// Returns the data inserted and optionally a verification error.
+    /// WriteLock is acquired only for the final update (after verification).
+    pub async fn insert(
+        self: Arc<Self>,
+        data: Vec<SignedAccountData>,
+    ) -> (Vec<SignedAccountData>, Option<Error>) {
+        let x = self.clone();
+        // Execute insertion in a dedicated runtime.
+        self.runtime
+            .spawn(async move {
```

Comment:
Are you sure that it is a good idea to have all CPU-heavy tasks on a single thread pool? Is there no threat of starvation in such setup?

Reply:
> dedicated threadpools feel like a work-around
> oversubscribing CPU also causes starvation, just in a different way.

Starvation is when you don't have a guarantee that a given task will eventually complete because it won't even obtain enough resources. A bounded number of threadpools (with bounded number of threads) with a round robin CPU assignment precisely gives a guarantee that no threadpool will starve.

> for known cases where we are spawning unbounded amount of background work, implement throttling, like we do here

This specific implementation of throttling under-utilizes CPU as in case there is no competing tasks we utilize only half of the CPUs. Also it is not so different than just assigning a dedicated threadpool with rayon::current_num_threads to the given task.

@matklad how is this rayon example really different than spawning exactly the same stuff on tokio, except that it doesn't support async waiting for results but rather makes you use extra channels?

@mm-near what is your thoughts on sharing single global thread pool for block processing and and other stuff? Didn't we have already issues of this sort?

Reply:
yes: 
https://github.com/matklad/nearcore/blob/5c582ce9bba19a9d74bf6858179c381d78647882/chain/chain/src/chain.rs#L1967-L1998